app/lib/validator_manager.py

```diff
@@ -36,8 +39,97 @@ def deregister_validators(ws_endpoint, sudo_seed, stash_account_addresses):
 
 
 def get_validators_to_add(ws_endpoint):
-    return substrate_query_url(ws_endpoint, "ValidatorManager", "ValidatorsToAdd")
+    # todo rename function from get_validators_to_add -> get_validators_pending
+    if network_consensus() == "poa":
+        return substrate_query_url(ws_endpoint, "ValidatorManager", "ValidatorsToAdd")
+    else:
+        active_validators = substrate_query_url(ws_endpoint, "Session", "Validators")
+        staking_validators = staking_validators_set(ws_endpoint)
+        return list(set(staking_validators) - set(active_validators))
 
 
 def get_validators_to_retire(ws_endpoint):
-    return substrate_query_url(ws_endpoint, "ValidatorManager", "ValidatorsToRetire")
+    # todo rename function from get_validators_to_retire -> get_validators_to_remove
+    if network_consensus() == "poa":
+        return substrate_query_url(ws_endpoint, "ValidatorManager", "ValidatorsToRetire")
+    else:
+        active_validators = substrate_query_url(ws_endpoint, "Session", "Validators")
+        staking_validators = staking_validators_set(ws_endpoint)
+        return list(set(active_validators) - set(staking_validators))
+
+
+def setup_pos_validator(ws_endpoint, session_key,  stash_seed, controller_seed=None):
+    batch_call = []
+    substrate_client = get_substrate_client(ws_endpoint)
+    stash_keypair = Keypair.create_from_uri(stash_seed)
+
+    batch_call.append(substrate_client.compose_call(
+            call_module='Staking',
+            call_function='bond',
+            call_params={
+                'controller': stash_keypair.ss58_address,
+                'value': 1 * 10 ** 11,
+                'payee': "Staked"
+            }
+        )
+    )
+
+    if type(session_key) == str:
+        session_key = decode_session_key(substrate_client,session_key)
```

Comment:
I will add the logic to skip.

Reply:
I checked the current logic of PoA validators and it doesn't have such logic, the key is rotated every time.
However I notice [logic](https://github.com/paritytech/testnet-manager/blob/main/app/lib/network_utils.py#L283) which prevents registering the active/pending/deleting nodes, I added it to PoS : https://github.com/paritytech/testnet-manager/pull/9/commits/c0af1dc83359057cc2e28337bfb01c1340145d89

I think the main logic can be refracted, since you may want to register back the "pending_deletion" node. but since it will affect the PoA and PoS validator it will be better if we split this. You may notice I already left several todo for refactoring: [L42](https://github.com/paritytech/testnet-manager/pull/9/files#diff-1f74e5848d18341b20fdfd2e7b7570b02e4ccbc8a07f5cd9f7335e1ec616d700R42) and [L52](https://github.com/paritytech/testnet-manager/pull/9/files#diff-1f74e5848d18341b20fdfd2e7b7570b02e4ccbc8a07f5cd9f7335e1ec616d700R52)

After this PR is merged I will create issue for refactoring with links to ToDo, I did not include refactoring in this PR since it will complicate the review process.